### PR TITLE
vendor: upgrade grpc from 1.13.0 to 1.21.2

### PR DESCRIPTION
**Pull request by ajwerner**
This PR upgrades gRPC from 1.13.0 to 1.21.2. The primary motivation for this
upgrade is to eliminate the disconnections caused by
https://github.com/grpc/grpc-go/issues/1882. These failures manifest themselves
as the following set of errors:

```
ajwerner-test-0001> I190722 22:15:01.203008 12054 vendor/github.com/cockroachdb/circuitbreaker/circuitbreaker.go:322  [n1] circuitbreaker: rpc [::]:26257 [n2] tripped: failed to check for ready connection to n2 at ajwerner-test-0002:26257: connection not ready: TRANSIENT_FAILURE
```
Which then lead to tripped breakers and general badness. I suspect that there
are several other good bug fixes in here, including some purported leaks and
correctness fixes on shutdown.

I have verified that with this upgrade I no longer see connections break in
overload scenarios which reliably reproduced the situation in the above log.

This commit removes one condition from grpcutil.IsClosedConnection which should
be subsumed by the status check above. The `transport` subpackage has not been
around for many releases.

This does not upgrade to the current release 1.22.0 because the maintainer
mentions that it contains a bug
(https://github.com/grpc/grpc-go/issues/2663#issuecomment-504129002).

Release note (bug fix): Upgrade grpc library to fix connection state management
bug.

### CODE REVIEW

**Comment by ajwerner** [comment]
Grr the following PR https://github.com/grpc/grpc-go/pull/2642 breaks `TestGRPCKeepaliveFailureFailsInflightRPCs` which expects to be able to control the Keepalive duration. Here the gRPC spec is working against us. I've updated the test to not time out until 30s but it makes it a 10s test. I'm open to skipping it. 

This test however under stress hit https://github.com/grpc/grpc-go/pull/2818 which is only fixed in 1.22 which is unfortunately subject to https://github.com/grpc/grpc-go/pull/2862.

In summary there's not published release which is thought to be bug free. I'm tempted to pull master as it purports to contain bug fixes for all of these bugs and then to watch for the next release. Thoughts?

**Comment by jordanlewis** [comment]
I don't feel particularly great about pulling master. What other bugs might be lurking?

What's the version released before 2642?

**Comment by ajwerner** [comment]
1.18.1, I've verified that it passes, though seems somewhat flakey under stress on a different test. Digging in to that now. 

Master also seems to fail the test unreliably due to what appears to be requests waiting on writeQuota. I can't tell if it's working as expected but it feels like a bug. The connection should be closed. I'll dig a little deeper and potentially file a grpc bug.

**Comment by ajwerner** [comment]
I saw this after 4 minutes of stress:

```
--- FAIL: TestStatsHandlerWithHeartbeats (45.10s)
    soon.go:35: condition failed to evaluate within 45s: expected server.incoming > client.outgoing; got 110, 134   
```

Digging. Otherwise this feels a bit better. Will also verify the disconnection behavior.

**Comment by ajwerner** [comment]
There's no winning here. 1.18.1 contains https://github.com/grpc/grpc-go/issues/2692 which causes the flakiness in TestStatsHandlerWithHeartbeats and also would just totally in general muck with our stats. 

I think I'm going to go back to trying to pick up 1.21.1 and adjust the tests appropriately.

**Comment by ajwerner** [comment]
Okay this is updated to use 1.21.1 and roughly RFAL.

This bump brings us to really long gRPC keepalives. It will mean that detecting network partitions from the client to the server will take ~3x longer but that's probably okay. The more annoying thing is that the test now takes much longer. The test also needed to be updated because it used to rely on having enough quota in flow control to send without blocking. It passes now reliably but I added a second commit to skip it because of its crazy long duration.

Another issue this brings up are that the stats in master (since 1.17) are totally broken (see https://github.com/grpc/grpc-go/pull/2932). This process has been rather frustrating. I'm wondering if it's worth having a fork of grpc for little things like this. I can try to get this change backported but the gRPC maintainers seem to have a tendency to let PRs sit for a long time.

**Comment by ajwerner** [comment]
Okay, good news, I got https://github.com/grpc/grpc-go/pull/2932 merged, backported, and released as 1.21.2. I also defeated the linter. This is now truly RFAL. @jordanlewis if you want to pass this off to another reviewer feel free.

**Comment by ajwerner** [comment]
@bdarnell or @asubiotto interested in having a look?

**Review by asubiotto — APPROVED** [review]
:lgtm: this seems like it closes #31361, right?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39041#-:-LkyMsYb4vAn0ko6cCK1:beqimcj)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

**Review by bdarnell — APPROVED** [review]
> Reviewed 10 of 11 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39041)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis)

**Comment by ajwerner** [comment]
TFTR! I'm going to close #31361 in favor of #34229 and #38602.
